### ISSUE TITLE
Field assigned with non-nullable `out` in constructor should count as initialized

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#41296]
The following should not produce any warning.

[update jcouv: fixed repro to actually repro the issue I meant]

``` csharp
#nullable enable

using System;
public class C 
{
    string field; // warning CS8618: Non-nullable field 'field' is uninitialized. Consider declaring the field as nullable.
    
    public C() 
    {
        try 
        {
            M2(out field); 
        }
        catch (Exception)
        {
           if (field is null)
           {
              field  =  "";   
           }
        }
    }
    
    static void M2(out string s) => throw null!;
}
```

This was encountered while annotating `PEFieldSymbol`.

FYI @sharwell @cston

**Comment by cston** [comment]
The warning looks correct - `field` is not initialized in the (default) parameter-less constructor.

**Comment by sharwell** [comment]
@cston but it is initialized. If `field` was a local variable declared at the start of the constructor, it would be considered definitely assigned at the end of the constructor due to its use as an `out` variable.

**Comment by cston** [comment]
There is no constructor in the example above. If `M()` is changed to a constructor, there are no warnings.

**Comment by jcouv** [comment]
Updated OP with an actual repro. The problem occurs with a `try/catch`.

**Comment by RikkiGibson** [comment]
The limitation here is that `if (x is null) x = value;` does not definitely assign `x`. It does update its state to not-null, though, so the new analysis will handle this properly.
